File: 3rd Semester/Advanced Programming Methods/Laboratory/My Toy Interpreter/src/domain/statements/FileOperations/ReadFileStmt.java
package domain.statements.FileOperations;

import domain.exp.Exp;
import domain.PrgState;
import domain.adt.Tuple;
import domain.statements.IStmt;
import exception.*;

import java.io.BufferedReader;
import java.io.IOException;

public class ReadFileStmt implements IStmt {
    private Exp exp;
    private String var;

    public ReadFileStmt(Exp exp, String var) {
        this.exp = exp;
        this.var = var;
    }

    @Override
    public PrgState execute(PrgState state) throws IOException, FileAlreadyOpenedException, FileNotOpenedException, UnknownVariableException, DivideByZeroException, UnknownComparisonExpression {
        int fd = this.exp.eval(state.getSymTable(), state.getHeap());
        Tuple<String, BufferedReader> br = state.getFileTable().get(fd);
        if (br == null)
            throw new FileNotOpenedException("FileNotOpenedException at: " + this.toString() + "\nNo such file descriptor: " + String.valueOf(fd));
        String line = br.getSecond().readLine();
        int val = 0;
        if(line != null)
            val = Integer.valueOf(line);
        state.getSymTable().put(this.var, val);
        return null;
    }

    @Override
    public String toString() {
        return "readFileStmt(" + this.exp.toString() + ", " + this.var + ")";
    }
}

File: 3rd Semester/Advanced Programming Methods/Laboratory/Toy Interpreter C#/sharplab/Model/Statements/readFileStmt.cs
using sharplab.Model.Expressions;
using sharplab.Model.Utils;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace sharplab.Model.Statements
{
    class ReadFileStmt : IStmt
    {
        private Exp exp;
        private String name;

        public ReadFileStmt(Exp exp, String name)
        {
            this.exp = exp;
            this.name = name;
        }

        public PrgState Execute(PrgState state)
        {
            IMyDictionary<string, int> symtab = state.GetSymTable();
            IFileTable<Utils.Tuple> tpl = state.GetFT();
       
            try{
                int val = exp.Eval(symtab);
                Utils.Tuple tup = tpl.Get(val);
                if (tup != null)
                {
                    StreamReader s = tup.GetStream();
                    String st = s.ReadLine();
                    if(st != null)
                        if(symtab.IsDefined(name)==0)
                            symtab.Put(name, int.Parse(st));
                        else
                            symtab.Put(name, int.Parse(st));
                    else
                        symtab.Put(name, 0);
                }
                else
                    throw new Exception("File does not exist!");
         
            }catch (IOException e)
            {
                System.Console.WriteLine(e.ToString());
            }
            return state;
        }

        public override string ToString()
        {
            return "Reading from: " + this.exp.ToString();
        }
    }
}
